### PR TITLE
Calculate storage bytes through interface method for encoding types.

### PR DESCRIPTION
**Pull request by hanhanW**
The revision moves the implementation of storage bytes calculation to `EncodingAttr::calculateStorageSizeInBytes`. If the encoding attribute implements the interface, the implementation has higher priority. Because it knows all the details, including whether packing the data back-to-back or not.

The change is not NFC because it also fixes a bug for dynamic cases. The `dynamicDims` value range is not a mixed value range. It is only for dynamic cases. To make the logic correct, we need to use `getDynamicDimIndex()` to get the corresponding dimension index before the update.

The revision duplicates two methods from Util to Encoding dialect because we do not want the dependency (i.e., encoding -> util):
  - getTypeBitWidth
  - getRoundedElementByteWidth

The function argument of `calculateStorageSizeInBytes` method is changed because of the needs.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Stream/Transforms/test/encode_host_tensors.mlir
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Stream/Transforms/test/encode_host_tensors.mlir]: This simplifies the computation (e.g., 64=16*4), but it adds overheads for the human. Do we want to use affine.apply ops in this case?

**Comment by hanhanW** [comment]
@lialan we can follow this to implement the packed i1 support. The stream side should work if you implement the interface method for your packed encoding type. Please take a look.

File: compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp]: This line fixes the bug that I mentioned in the PR description.

**Comment by benvanik** [comment]
We should run some compile time profiles with this on large dynamically shaped programs - when I tried to use affine ops for this previously it caused a massive regression.

File: compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp]: We don't want this dependency (encoding -> util) - util only has those functions because an encoding is not specified - this dialect should precisely specify the bit widths without needing to fall back to the auto detection behavior.

That is, it'd make more sense for the util functions to check if the type has an encoding and ask the encoding for the size instead of the encoding asking util to guess the size for something it should know. This may mean a bit of duplication of the util logic into here but it's simple and will be something we want to override for various encodings (block based encodings won't use that logic, for example, and we never want to have this code path calling into the generic util guessing code).

**Review by benvanik — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp]: Thanks for taking a look! I do have this dependency question. :)
I saw both solution, but I did not know which one is better. The logic should be simple, so duplication looks okay to me. Good to know that there is no such concern.

**Comment by hanhanW** [comment]
> We should run some compile time profiles with this on large dynamically shaped programs - when I tried to use affine ops for this previously it caused a massive regression.

I see. I do not have a large dynamically programs on hands. So I'll go with arith ops for now, and we can revisit it later.

File: compiler/src/iree/compiler/Utils/ElementPackingUtils.cpp
Comment (lialan) [inline compiler/src/iree/compiler/Utils/ElementPackingUtils.cpp:85]: After the change, are we able to guarantee here that we don't need to pad shape?

Comment (hanhanW) [reply compiler/src/iree/compiler/Utils/ElementPackingUtils.cpp:85]: No, we don't need to pad the shape anymore. The padding was required because of the encoding. I.e., this commit is an improvement of https://github.com/iree-org/iree/commit/e8f4948d5c62beea27a109161b697d9088d5b3bd#diff-9ccfc5b598b4e78bd96d52bb2d98f6a07d2d4c34803f80fe3a15248c8700b7a0R64

Instead of implementing the logic in the method, we now move the logic to the interface.

**Review by lialan — APPROVED** [review]

**Comment by lialan** [comment]
awesome. I will rebase my i1 attr branch https://github.com/iree-org/iree/tree/lialan/i1_attr on top of it once it is landed

**Comment by hanhanW** [comment]
@benvanik can I get a review from you? Thanks!

**Review by benvanik — APPROVED** [review]

**Pull request merged**
